File: frontend/src/api/wallet.ts
import { apiClient } from './client';
import type { ApiResponse, ArtifactTransaction } from '@/types';

export interface BalanceItem {
  artifact_type: string;
  label: string;
  quantity: number;
  usd_value: number;
}

export interface BalanceResponse {
  balance: BalanceItem[];
  total_label: string;
  total_fiat: number;
  fiat_currency: string;
  regular_balance: BalanceItem[];
  regular_total_fiat: number;
  creator_balance: BalanceItem[];
  creator_total_fiat: number;
  creator_display_name: string;
}

export interface BundleInfo {
  id: string;
  artifact_type: string;
  artifact_label: string;
  quantity: number;
  price_usd: number;
  savings: number;
}

export interface InitializeResponse {
  tx_ref: string;
  flutterwave_ref?: string;
  status?: string;
  requires_otp?: boolean;
  public_key: string;
  amount?: number;
  currency?: string;
  customer_email?: string;
  customer_name?: string;
}

export interface BankInfo {
  code: string;
  name: string;
}

export const walletApi = {
  getBalance: () =>
    apiClient.get<ApiResponse<BalanceResponse>>('/wallet/balance/').then((r) => r.data),

  getTransactions: (params?: { type?: string; direction?: string; cursor?: string }) =>
    apiClient.get<ApiResponse<ArtifactTransaction[]>>('/wallet/transactions/', { params }).then((r) => r.data),

  initializePurchase: (artifact_type: string, quantity: number, payment_method: string, options?: { bundle?: string; mpesa_phone?: string; card_details?: Record<string, string> }) =>
    apiClient.post<ApiResponse<InitializeResponse>>('/wallet/purchase/initialize/', {
      artifact_type, quantity, payment_method, ...options,
    }).then((r) => r.data),

  confirmPurchase: (tx_ref: string, flutterwave_id: string) =>
    apiClient.post<ApiResponse<{ transaction: ArtifactTransaction; new_balance: Record<string, number> }>>('/wallet/purchase/confirm/', {
      tx_ref, flutterwave_id,
    }).then((r) => r.data),

  tip: (username: string, artifact_type: string, quantity: number, message?: string) =>
    apiClient.post<ApiResponse<null>>('/wallet/tip/', { username, artifact_type, quantity, message }).then((r) => r.data),

  gift: (username: string, artifact_type: string, quantity: number, message?: string) =>
    apiClient.post<ApiResponse<null>>('/wallet/gift/', { username, artifact_type, quantity, message }).then((r) => r.data),

  withdraw: (artifact_type: string, quantity: number, method: string, options?: { phone_number?: string; bank_account?: string; bank_code?: string; account_name?: string; source?: 'regular' | 'creator' }) =>
    apiClient.post<ApiResponse<ArtifactTransaction>>('/wallet/withdraw/', {
      artifact_type, quantity, method, ...options,
    }).then((r) => r.data),

  getBanks: (country?: string) =>
    apiClient.get<ApiResponse<BankInfo[]>>('/wallet/withdraw/banks/', { params: country ? { country } : {} }).then((r) => r.data),

  resolveBank: (account_number: string, bank_code: string) =>
    apiClient.post<ApiResponse<{ account_number: string; bank_code: string; account_name: string }>>('/wallet/withdraw/bank-resolve/', {
      account_number, bank_code,
    }).then((r) => r.data),

  getBundles: () =>
    apiClient.get<ApiResponse<BundleInfo[]>>('/wallet/bundles/').then((r) => r.data),

  getExchangeRates: () =>
    apiClient.get<ApiResponse<{ rates: Record<string, number>; base_currency: string; local_currency: string; conversion_rate: number; labels: Record<string, string> }>>('/wallet/exchange-rates/').then((r) => r.data),

  transferFromCreator: (artifact_type: string, quantity: number) =>
    apiClient.post<ApiResponse<{ regular_balance: Record<string, number>; creator_balance: Record<string, number> }>>('/wallet/creator/transfer/', {
      artifact_type, quantity,
    }).then((r) => r.data),

  updateCreatorProfile: (data: { creator_display_name?: string }) =>
    apiClient.patch<ApiResponse<{ creator_display_name: string; creator_balance: Record<string, number> }>>('/wallet/creator/profile/', data).then((r) => r.data),

  requestPayout: (data: { amount?: number; artifact_type?: string; quantity?: number; method: string; phone_number?: string; bank_account?: string; bank_code?: string; account_name?: string }) =>
    apiClient.post<ApiResponse<ArtifactTransaction>>('/wallet/payout-request/', data).then((r) => r.data),

  getPayoutHistory: () =>
    apiClient.get<ApiResponse<ArtifactTransaction[]>>('/wallet/payout-history/').then((r) => r.data),
};

File: buddy_up_flutter/lib/data/models/wallet.dart
import 'package:freezed_annotation/freezed_annotation.dart';

part 'wallet.freezed.dart';
part 'wallet.g.dart';

@freezed
abstract class BalanceItem with _$BalanceItem {
  const factory BalanceItem({
    @JsonKey(name: 'artifact_type') required String artifactType,
    required String label,
    required int quantity,
    @JsonKey(name: 'usd_value') required double usdValue,
  }) = _BalanceItem;

  factory BalanceItem.fromJson(Map<String, dynamic> json) =>
      _$BalanceItemFromJson(json);
}

@freezed
abstract class BalanceResponse with _$BalanceResponse {
  const factory BalanceResponse({
    required List<BalanceItem> balance,
    @JsonKey(name: 'total_label') required String totalLabel,
    @JsonKey(name: 'total_fiat') required double totalFiat,
    @JsonKey(name: 'fiat_currency') required String fiatCurrency,
    @JsonKey(name: 'regular_balance') @Default(<BalanceItem>[]) List<BalanceItem> regularBalance,
    @JsonKey(name: 'regular_total_fiat') @Default(0.0) double regularTotalFiat,
    @JsonKey(name: 'creator_balance') @Default(<BalanceItem>[]) List<BalanceItem> creatorBalance,
    @JsonKey(name: 'creator_total_fiat') @Default(0.0) double creatorTotalFiat,
    @JsonKey(name: 'creator_display_name') @Default('') String creatorDisplayName,
  }) = _BalanceResponse;

  factory BalanceResponse.fromJson(Map<String, dynamic> json) =>
      _$BalanceResponseFromJson(json);
}

@freezed
abstract class ArtifactTransaction with _$ArtifactTransaction {
  const factory ArtifactTransaction({
    required String id,
    @JsonKey(name: 'transaction_type') required String transactionType,
    @JsonKey(name: 'artifact_type') required String artifactType,
    required int quantity,
    required String direction,
    @JsonKey(name: 'counterparty_id') String? counterpartyId,
    @JsonKey(name: 'counterparty_name') String? counterpartyName,
    @JsonKey(name: 'reference_id') required String referenceId,
    required String status,
    @JsonKey(name: 'fiat_amount') String? fiatAmount,
    @JsonKey(name: 'fiat_currency') required String fiatCurrency,
    String? description,
    @JsonKey(name: 'clearance_at') String? clearanceAt,
    @JsonKey(name: 'created_at') required String createdAt,
  }) = _ArtifactTransaction;

  factory ArtifactTransaction.fromJson(Map<String, dynamic> json) =>
      _$ArtifactTransactionFromJson(json);
}

@freezed
abstract class BundleInfo with _$BundleInfo {
  const factory BundleInfo({
    required String id,
    @JsonKey(name: 'artifact_type') required String artifactType,
    @JsonKey(name: 'artifact_label') required String artifactLabel,
    required int quantity,
    @JsonKey(name: 'price_usd') required double priceUsd,
    required double savings,
  }) = _BundleInfo;

  factory BundleInfo.fromJson(Map<String, dynamic> json) =>
      _$BundleInfoFromJson(json);
}

@freezed
abstract class BankInfo with _$BankInfo {
  const factory BankInfo({
    required String code,
    required String name,
  }) = _BankInfo;

  factory BankInfo.fromJson(Map<String, dynamic> json) =>
      _$BankInfoFromJson(json);
}

@freezed
abstract class BankResolveResult with _$BankResolveResult {
  const factory BankResolveResult({
    @JsonKey(name: 'account_number') required String accountNumber,
    @JsonKey(name: 'bank_code') required String bankCode,
    @JsonKey(name: 'account_name') required String accountName,
  }) = _BankResolveResult;

  factory BankResolveResult.fromJson(Map<String, dynamic> json) =>
      _$BankResolveResultFromJson(json);
}

@freezed
abstract class InitializePurchaseResponse with _$InitializePurchaseResponse {
  const factory InitializePurchaseResponse({
    @JsonKey(name: 'tx_ref') required String txRef,
    @JsonKey(name: 'flutterwave_ref') String? flutterwaveRef,
    String? status,
    @JsonKey(name: 'requires_otp') @Default(false) bool requiresOtp,
    @JsonKey(name: 'public_key') required String publicKey,
    double? amount,
    String? currency,
    @JsonKey(name: 'customer_email') String? customerEmail,
    @JsonKey(name: 'customer_name') String? customerName,
  }) = _InitializePurchaseResponse;

  factory InitializePurchaseResponse.fromJson(Map<String, dynamic> json) =>
      _$InitializePurchaseResponseFromJson(json);
}

@freezed
abstract class ConfirmPurchaseResponse with _$ConfirmPurchaseResponse {
  const factory ConfirmPurchaseResponse({
    required ArtifactTransaction transaction,
    @JsonKey(name: 'new_balance') required Map<String, int> newBalance,
  }) = _ConfirmPurchaseResponse;

  factory ConfirmPurchaseResponse.fromJson(Map<String, dynamic> json) =>
      _$ConfirmPurchaseResponseFromJson(json);
}

@freezed
abstract class ExchangeRates with _$ExchangeRates {
  const factory ExchangeRates({
    required Map<String, double> rates,
    @JsonKey(name: 'base_currency') required String baseCurrency,
    @JsonKey(name: 'local_currency') required String localCurrency,
    @JsonKey(name: 'conversion_rate') required double conversionRate,
    required Map<String, String> labels,
  }) = _ExchangeRates;

  factory ExchangeRates.fromJson(Map<String, dynamic> json) =>
      _$ExchangeRatesFromJson(json);
}
